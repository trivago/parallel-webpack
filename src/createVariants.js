var assign = require('lodash.assign');
var flatten = require('lodash.flatten');

/**
 * Creates configuration variants.
 *
 * @param {Object} [baseConfig={}] The base configuration
 * @param {Object} variants The variants
 * @param {Function} [configCallback] A callback executed for each configuration to
 *      transform the variant into a webpack configuration
 * @returns {*|Array}
 */
module.exports = function createVariants(baseConfig, variants, configCallback) {
    if(arguments.length < 3) {
        if(arguments.length === 2) {
            if(typeof variants === 'function') {
                // createVariants(variants: Object, configCallback: Function)
                configCallback = variants;
                variants = baseConfig;
                baseConfig = {};
            }
            // createVariants(baseConfig: Object, variants: Object)
            // => don't do anything
        } else {
            // createVariants(variants: Object)
            variants = baseConfig;
            baseConfig = {};
        }
    }

    // Okay, so this looks a little bit messy but it really does make some sense.
    // Essentially, for each base configuration, we want to create every
    // possible combination of the configuration variants specified above.
    var transforms = Object.keys(variants).map(function(key) {
            return function(config) {
                return variants[key].map(function(value) {
                    var result = assign({}, config);
                    result[key] = value;
                    return result;
                });
            };
        }),
        configs = transforms.reduce(function(options, transform) {
            return flatten(options.map(transform));
        }, [baseConfig]);


    return configCallback && configs.map(configCallback) || configs;
};
