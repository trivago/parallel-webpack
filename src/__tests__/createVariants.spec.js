import createVariants from '../createVariants';

describe('createVariants', () => {
    function test(withBase, withCallback) {
        // these are stupid examples but enough to prove transformation works
        const baseConfig = {
            name: 'bundle.base'
        };
        const variants = {
            rtl: [true, false],
            debug: [true, false],
            features: [0, 1, 2]
        };
        const callback = (x) => {
            const ret = Object.assign({}, x);
            ret.name = `bundle.${x.rtl ? 'rtl' : 'no-rtl'}.${x.debug ? 'dbg': 'no-debug'}.${x.features}.js`;
            return ret;
        };

        let result;

        if (withBase && withCallback) {
            result = createVariants(baseConfig, variants, callback);
        } else if (withBase && !withCallback) {
            result = createVariants(baseConfig, variants);
        } else if (!withBase && withCallback) {
            result = createVariants(variants, callback);
        } else { //!withBase && !withCallback
            result = createVariants(variants);
        }
        expect(result).toMatchSnapshot();
    }
    it('should generate configs with "baseConfig, variants, configCallback" as params', () => {
        test(true, true);
    });
    it('should generate configs with "variants, configCallback" as params', () => {
        test(false, true);
    });
    it('should generate configs with "variants" as params', () => {
        test(false, false);
    });

    it('should generate configs with baseConfig and variants', () => {
        test(true, false);
    });
});
