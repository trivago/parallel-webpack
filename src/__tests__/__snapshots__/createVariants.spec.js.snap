// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createVariants should generate configs with "baseConfig, variants, configCallback" as params 1`] = `
Array [
  Object {
    "debug": true,
    "features": 0,
    "name": "bundle.rtl.dbg.0.js",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 1,
    "name": "bundle.rtl.dbg.1.js",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 2,
    "name": "bundle.rtl.dbg.2.js",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 0,
    "name": "bundle.rtl.no-debug.0.js",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 1,
    "name": "bundle.rtl.no-debug.1.js",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 2,
    "name": "bundle.rtl.no-debug.2.js",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 0,
    "name": "bundle.no-rtl.dbg.0.js",
    "rtl": false,
  },
  Object {
    "debug": true,
    "features": 1,
    "name": "bundle.no-rtl.dbg.1.js",
    "rtl": false,
  },
  Object {
    "debug": true,
    "features": 2,
    "name": "bundle.no-rtl.dbg.2.js",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 0,
    "name": "bundle.no-rtl.no-debug.0.js",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 1,
    "name": "bundle.no-rtl.no-debug.1.js",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 2,
    "name": "bundle.no-rtl.no-debug.2.js",
    "rtl": false,
  },
]
`;

exports[`createVariants should generate configs with "variants" as params 1`] = `
Array [
  Object {
    "debug": true,
    "features": 0,
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 1,
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 2,
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 0,
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 1,
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 2,
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 0,
    "rtl": false,
  },
  Object {
    "debug": true,
    "features": 1,
    "rtl": false,
  },
  Object {
    "debug": true,
    "features": 2,
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 0,
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 1,
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 2,
    "rtl": false,
  },
]
`;

exports[`createVariants should generate configs with "variants, configCallback" as params 1`] = `
Array [
  Object {
    "debug": true,
    "features": 0,
    "name": "bundle.rtl.dbg.0.js",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 1,
    "name": "bundle.rtl.dbg.1.js",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 2,
    "name": "bundle.rtl.dbg.2.js",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 0,
    "name": "bundle.rtl.no-debug.0.js",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 1,
    "name": "bundle.rtl.no-debug.1.js",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 2,
    "name": "bundle.rtl.no-debug.2.js",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 0,
    "name": "bundle.no-rtl.dbg.0.js",
    "rtl": false,
  },
  Object {
    "debug": true,
    "features": 1,
    "name": "bundle.no-rtl.dbg.1.js",
    "rtl": false,
  },
  Object {
    "debug": true,
    "features": 2,
    "name": "bundle.no-rtl.dbg.2.js",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 0,
    "name": "bundle.no-rtl.no-debug.0.js",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 1,
    "name": "bundle.no-rtl.no-debug.1.js",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 2,
    "name": "bundle.no-rtl.no-debug.2.js",
    "rtl": false,
  },
]
`;

exports[`createVariants should generate configs with baseConfig and variants 1`] = `
Array [
  Object {
    "debug": true,
    "features": 0,
    "name": "bundle.base",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 1,
    "name": "bundle.base",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 2,
    "name": "bundle.base",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 0,
    "name": "bundle.base",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 1,
    "name": "bundle.base",
    "rtl": true,
  },
  Object {
    "debug": false,
    "features": 2,
    "name": "bundle.base",
    "rtl": true,
  },
  Object {
    "debug": true,
    "features": 0,
    "name": "bundle.base",
    "rtl": false,
  },
  Object {
    "debug": true,
    "features": 1,
    "name": "bundle.base",
    "rtl": false,
  },
  Object {
    "debug": true,
    "features": 2,
    "name": "bundle.base",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 0,
    "name": "bundle.base",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 1,
    "name": "bundle.base",
    "rtl": false,
  },
  Object {
    "debug": false,
    "features": 2,
    "name": "bundle.base",
    "rtl": false,
  },
]
`;
